File: java/src/graphs/flows/MaxFlowFordFulkersonSimple.java
package graphs.flows;

// https://en.wikipedia.org/wiki/Ford–Fulkerson_algorithm in O(V^2 * flow)
public class MaxFlowFordFulkersonSimple {

    public static int maxFlow(int[][] cap, int s, int t) {
        for (int flow = 0; ; ++flow)
            if (!augmentPath(cap, new boolean[cap.length], s, t))
                return flow;
    }

    static boolean augmentPath(int[][] cap, boolean[] vis, int i, int t) {
        if (i == t)
            return true;
        vis[i] = true;
        for (int j = 0; j < vis.length; j++)
            if (!vis[j] && cap[i][j] > 0 && augmentPath(cap, vis, j, t)) {
                --cap[i][j];
                ++cap[j][i];
                return true;
            }
        return false;
    }

    // Usage example
    public static void main(String[] args) {
        int[][] capacity = {{0, 1, 1, 0}, {1, 0, 1, 1}, {1, 1, 0, 1}, {0, 1, 1, 0}};
        System.out.println(2 == maxFlow(capacity, 0, 3));
    }
}

File: python/maxflow_fordfulkerson_simple.py
# https://en.wikipedia.org/wiki/Ford–Fulkerson_algorithm in O(V^2 * flow)


def max_flow(cap, s, t):
    def augment_path(cap, visited, i, t):
        if i == t:
            return True
        visited[i] = True
        for j in range(len(cap)):
            if not visited[j] and cap[i][j] > 0 and augment_path(cap, visited, j, t):
                cap[i][j] -= 1
                cap[j][i] += 1
                return True
        return False

    flow = 0
    while True:
        if not augment_path(cap, [False] * len(cap), s, t):
            return flow
        flow += 1


def test():
    capacity = [[0, 1, 1, 0], [1, 0, 1, 1], [1, 1, 0, 1], [0, 1, 1, 0]]
    assert max_flow(capacity, 0, 3) == 2


test()
